backend_plugin/src/main/evaluators/CorrectIndentationEvaluator.java

```diff
@@ -0,0 +1,131 @@
+     * of code in a string to reference later
+     */
+    @Override
+    public boolean evaluateDocumentBeforeChange(DocumentEvent event) {
+	// We only care about adding white space or removing characters
+	// Either of those cases will have a length of 0 once the string is trimmed
+	if (event.getText().trim().length() == 0) {
+	    try {
+		int line = document.getLineOfOffset(event.getOffset());
+		int lineOffset = document.getLineOffset(line);
+		int lineLength = document.getLineLength(line);
+		lineBeforeChange = document.get(lineOffset, lineLength);
+		whiteSpaceAddedOrRemoved = true;
+		return false;
+	    } catch (BadLocationException e) {}
+	}
+	whiteSpaceAddedOrRemoved = false;
+	return false;
+    }
+
+    /**
+     * If the change added white space or removed some character then compare the line of code before the change
+     * and after the change to determine if whitespace was added or removed at the start of the line
+     */
+    @Override
+    public boolean evaluateDocumentChanges(DocumentEvent event) {
+	try {
+	    // Since we already checked that we are only adding whitespace or removing text, we just check the boolean
+	    if (whiteSpaceAddedOrRemoved) {
+		int line = document.getLineOfOffset(event.getOffset());
+		if (lineIndentationChanged(event, line) && lineAdjacentToPrevious(line)) {
+		    return true;
+		}
+	    }
+	} catch (BadLocationException e) {}
+	return false;
+    }
+
+    /**
+     * Checks that the only change to the document line was the addition or removal of white space
+     * @param event the DocumentEvent that occurred
+     * @param line the line number where the change occurred
+     * @return true if white space was added or removed to the front of the line, false otherwise
+     */
+    private boolean lineIndentationChanged(DocumentEvent event, int line) {
+	try {
+	    // if both before and after are identical after trim
+	    // and the starting character is different, then we have a changed indentation
+	    int lineOffset = document.getLineOffset(line);
+	    int lineLength = document.getLineLength(line);
+	    String lineAfterChange = document.get(lineOffset, lineLength);
+
+	    if (lineAfterChange.trim().length() == lineBeforeChange.trim().length() &&
+		    (getOffsetOfFirstCharInLine(lineAfterChange) != getOffsetOfFirstCharInLine(lineBeforeChange))) {
+		return true;
+	    }
+	} catch (BadLocationException e) {}
+	return false;
+    }
+
+    /**
+     * Returns the position of the first non-white space character in the provided string
+     * @param text the string to find the first non-white space character in
+     * @return the position of the first non-white space character, or -2 if no such character was found
+     */
+    private int getOffsetOfFirstCharInLine(String text) {
+	// iterate through until we hit something that's not " " or "\t"
+	for (int i = 0; i < text.length(); i++) {
+	    char c = text.charAt(i);
+	    if (c != ' ' && c != '\t') {
```

Comment:
You can use `Character.isWhitespace(text.charAt(i))` here.

Reply:
Great catch. I had no idea this even existed.